include/verilated_vpi.cpp

```diff
@@ -589,11 +615,50 @@ class VerilatedVpiImp final {
         }
     }
     static QData cbNextDeadline() {
-        const auto it = s().m_timedCbs.cbegin();
-        if (VL_LIKELY(it != s().m_timedCbs.cend())) return it->first.first;
-        return ~0ULL;  // maxquad
+        QData min = ~0ULL;
+        auto it = s().m_timedCbs.cbegin();
+        if (VL_LIKELY(it != s().m_timedCbs.cend())) min = std::min(min, it->first.first);
+        it = s().m_readWriteCbs.cbegin();
+        if (VL_LIKELY(it != s().m_readWriteCbs.cend())) min = std::min(min, it->first.first);
+        it = s().m_readOnlyCbs.cbegin();
+        if (VL_LIKELY(it != s().m_readOnlyCbs.cend())) min = std::min(min, it->first.first);
+        it = s().m_nextSimTimeCbs.cbegin();
+        if (VL_LIKELY(it != s().m_nextSimTimeCbs.cend())) min = std::min(min, it->first.first);
+        return min;
+    }
+    static bool callTimeCbs(const uint32_t reason) VL_MT_UNSAFE_ONE {
+        const QData time = VL_TIME_Q();
+        std::map<std::pair<QData, uint64_t>, VerilatedVpiCbHolder>* vec;
+        bool called = false;
+        if (reason == cbNextSimTime) {
+            vec = &s().m_nextSimTimeCbs;
+        } else if (reason == cbReadWriteSynch) {
+            vec = &s().m_readWriteCbs;
+        } else {
+            vec = &s().m_readOnlyCbs;
+        }
```

Comment:
Should we instead check for cbReadOnly and assert(0) if not?  (I understand upper code will only call in these cases, but we might change that someday.)